lib/etl/initializer.ex

```diff
@@ -0,0 +1,74 @@
+defmodule Etl.Initializer do
+  defmodule Stage do
+    defstruct [:child_spec, :subscription_opts]
+  end
+
+  def start(stages, context) do
+    %{pids: pids, subs: subs} = start_pipeline(stages, context, %{pids: [], subs: []})
+
+    {Enum.reverse(pids), Enum.reverse(subs)}
+  end
+
+  defp start_pipeline([], _context, result), do: result
+
+  defp start_pipeline([%Stage{} = stage | remaining], context, %{pids: pids, subs: subs}) do
+    {:ok, pid} = start_child(stage.child_spec, context)
+
+    {:ok, sub} = GenStage.sync_subscribe(pid, stage.subscription_opts)
+
+    next(remaining, pid, context, %{pids: [pid | pids], subs: [sub | subs]})
+  end
+
+  defp start_pipeline([child_spec | remaining], context, %{pids: pids, subs: subs}) do
+    {:ok, pid} = start_child(child_spec, context)
+
+    next(remaining, pid, context, %{pids: [pid | pids], subs: subs})
+  end
+
+  defp next([], _pid, _context, result) do
+    result
+  end
+
+  defp next([child_spec | remaining], pid, context, result) do
```

Comment:
This is clever and hard to understand (as any clever solution should be). I think you need a better function name than `next`. How about `continue_pipeline_maybe_partitions`? Or `maybe_partition_pipeline`?

Reply:
Fair but there’s no maybe about it; _something_ is always triggered here unless it’s the end of the pipeline. What about `setup_next_stage` or some such?

Reply:
You always continue the pipeline. Maybe with partition stages, maybe not. But perhaps it's not worth calling out in the function name.

I would go with `continue_pipeline` then, since the entry function is `start_pipeline`.

Reply:
I see what you're saying; i'm focusing on the result of the function (the next stage in the pipeline is wired up for starting) and you're thinking about what the function is doing (wiring up one but possibly more stages). For me, the critical piece of the `next` function is that it takes the next link in the chain and casts any of its partitions to a `%Stage{}` struct.

Reading it again, you could make the argument both functions could be renamed to something like `start_and_sync_stage` instead of `start_pipeline` and `prepare_next_stage` instead of `next`, but again that's really with an eye toward the output of the function, not what it's doing to get there.